### DISCUSSION TITLE
Jotai-first way of updating data via listeners

### DISCUSSION THREAD

**Discussion by tomhicks** [pmndrs/jotai#271] [Q&A] [answered]
Hi,

I've got an app where I am using Firebase Realtime Database (RTDB). The data in these updates via websockets, and you receive change events to keep your local state in sync. For "collections", you get events whenever an item is added, changed or removed.

At the moment, I have a lot of hooks like `useSyncPeopleToAtom` or `useSyncFramesToAtom` - like this:
```
const framesAtom = atom<Record<string, Frame>>({})
export function useSyncFramesToAtom(): void {
  const documentId = useDocumentId() // from context (but could be atomised if it helps)
  const setFrames = useUpdateAtom(framesAtom)

  useEffect(() => {
    const framesListRef = getCollectionRefRef(documentId)

    framesListRef.on("child_added", (r) => {
      const id = r.key
      if (!id) return
      setFrames(assoc(id, {id, ...r.val()}))
    })

    framesListRef.on("child_removed", (r) => {
      const id = r.key
      if (!id) return
      setFrames(omit([id]))
    })

    framesListRef.on("child_changed", (r) => {
      const id = r.key
      if (!id) return
      setFrames(assoc(id, {id, ...r.val()}))
    })

    return () => {
      framesListRef.off("child_added")
      framesListRef.off("child_changed")
      framesListRef.off("child_removed")
      setFrames({})
    }
  }, [documentId, setFrames])
}
```

I'm basically doing a whole load of these at the top of my "document" view, in order to sync lots of different parts from the document.

I'm wondering if there is an "atomised" way of doing this, where I can set up this updating mechanism and then when I request to use that atom, it's set up, rather than having this big list of `useSyncCollectionXToAtom, useSyncCollectionYToAtom, useSyncCollectionZToAtom` in my top-level component?

(Oh, and I'm also going to begin to use `atomFamilies` soon, in case that affects things)

**Comment by dai-shi** [comment, marked as answer]
Your current approach is correct and valid in the current version of jotai.
There has been some discussions in #211, #234, #263, and it's blocking #248.
So, we started working on #266.

You are welcome to join the discussion there, help on ideation, testing, documenting, and creating examples.
Especially, because I haven't thought much about how to apply this to `atomFamily`, your feedback would be so helpful.
